BedrockConflictMetrics.cpp

```diff
@@ -0,0 +1,103 @@
+#include "BedrockConflictMetrics.h"
+
+// Initialize non-const static variables.
+recursive_mutex BedrockConflictMetrics::_mutex;
+map<string, BedrockConflictMetrics> BedrockConflictMetrics::_conflictInfoMap;
+
+BedrockConflictMetrics::BedrockConflictMetrics(const string& commandName) :
+_commandName(commandName)
+{ }
+
+void BedrockConflictMetrics::success() {
+    ++_totalSuccessCount;
+    _results.reset(_resultsPtr);
+    ++_resultsPtr;
+    _resultsPtr %= COMMAND_COUNT;
+}
+
+void BedrockConflictMetrics::conflict() {
+    ++_totalConflictCount;
+    _results.set(_resultsPtr);
+    ++_resultsPtr;
+    _resultsPtr %= COMMAND_COUNT;
+}
+
+int BedrockConflictMetrics::recentSuccessCount() {
+    return COMMAND_COUNT - _results.count();
+}
+
+int BedrockConflictMetrics::recentConflictCount() {
+    return _results.count();
+}
+
+void BedrockConflictMetrics::recordConflict(const string& commandName) {
+    SAUTOLOCK(_mutex);
+
+    // Look up, and create if necessary, the info object for this command.
+    auto it = _conflictInfoMap.find(commandName);
+    if (it == _conflictInfoMap.end()) {
+        // We need to insert this command.
+        auto itPair = _conflictInfoMap.emplace(make_pair(commandName, commandName));
+        it = itPair.first;
+    }
+    it->second.conflict();
+    SINFO("Multi-write conflict recorded for " << commandName);
+}
+
+void BedrockConflictMetrics::recordSuccess(const string& commandName) {
+    SAUTOLOCK(_mutex);
+
+    // Look up, and create if necessary, the info object for this command.
+    auto it = _conflictInfoMap.find(commandName);
+    if (it == _conflictInfoMap.end()) {
+        // We need to insert this command.
+        auto itPair = _conflictInfoMap.emplace(make_pair(commandName, commandName));
+        it = itPair.first;
+    }
+    it->second.success();
+    SINFO("Multi-write success recorded for " << commandName);
+}
+
+bool BedrockConflictMetrics::multiWriteOK(const string& commandName) {
+    SAUTOLOCK(_mutex);
+    bool result;
+
+    // Look up the command in the list of commands.
+    auto it = _conflictInfoMap.find(commandName);
+    if (it == _conflictInfoMap.end()) {
+        // If we don't find it, we assume it's OK to multi-write.
+        SINFO("Multi-write command '" << commandName << "' not tracked in BedrockConflictMetrics. Assuming OK.");
+        result = true;
+    } else {
+        // Otherwise, we check to see if it's recent conflict count is too high for multi-write.
+        auto& metric = it->second;
+        int conflicts = metric.recentConflictCount();
+        uint64_t totalAttempts = metric._totalConflictCount + metric._totalSuccessCount;
```

Comment:
The bit vector *is* initialized to all success, we don't need `_totalSuccessCount`, but I thought it was interesting information to log. Notably, `_totalSuccessCount` isn't just a count of successes in the last 100 tries, but in total since we initialized this command. We use it so that when we turn a command on/off for multi-write, we'll log the number of successes and failures for the life of the process. I.e., you can get a log line that looks like:

```
Multi-write changing to DENIED for command 'idcollision' recent conflicts: 10/100, total conflicts: 26/188
```

`totalSuccessCount` is just for that `188` number.